### PR TITLE
64550 Add 'at_quick_exit' to 10.14 whitelist

### PR DESCRIPTION
**Pull request by JDLH**
Apple clang version 11.0.0 (clang-1100.0.33.12) appears to lack the C11 language feature 'at_quick_exit' when compiling for macOS X 10.14. This causes the compiler to generate a `-Wimplicit-function-declaration` error message, even for correct configuration code, which correctly has a `#include <stdlib.h>` directive. MacPorts elevates this to the main.log. But it is a false positive. The configuration code is not implicitly declaring 'at_quick_exit'. So, this function name belongs in the whitelist.

An example of configuration code which demonstrates this condition is:

	#include <stdlib.h>
	static void func(void)
	{}

	int
	main (void)
	{
	 at_quick_exit(func);
	  ;
	  return 0;
	}

The verbose compiler output from my test of this code on macOS X 10.14.6 Mojave, using the macOS X 10.14 SDK, is:

	% /usr/bin/clang -v -o conftest -std=c11 -Werror -Wall -Wpointer-arith -Wcast-align -Wno-tautological-compare  conftest.c  -liconv                                                                
	Apple clang version 11.0.0 (clang-1100.0.33.12)
	Target: x86_64-apple-darwin18.7.0
	Thread model: posix
	InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
	 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" -cc1 -triple x86_64-apple-macosx10.14.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -mrelax-all -disable-free -disable-llvm-verifier -discard-value-names -main-file-name conftest.c -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -fno-strict-return -masm-verbose -munwind-tables -target-sdk-version=10.14 -target-cpu penryn -dwarf-column-info -debugger-tuning=lldb -ggnu-pubnames -target-linker-version 520 -v -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.0 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -Werror -Wall -Wpointer-arith -Wcast-align -Wno-tautological-compare -Wno-framework-include-private-from-public -Wno-atimport-in-framework-header -Wno-extra-semi-stmt -Wno-quoted-include-in-framework-header -std=c11 -fdebug-compilation-dir /Users/jdlh/workspace/freeciv_overall/freeciv_S2_6 -ferror-limit 19 -fmessage-length 132 -stack-protector 1 -mdarwin-stkchk-strong-link -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fobjc-runtime=macosx-10.14.0 -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -o /var/folders/5p/zmlrjq6c8xj84s0059bnypz80000gp/T/conftest-901715.o -x c conftest.c
	clang -cc1 version 11.0.0 (clang-1100.0.33.12) default target x86_64-apple-darwin18.7.0
	ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/local/include"
	ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks"
	#include "..." search starts here:
	#include <...> search starts here:
	 /usr/local/include
	 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/11.0.0/include
	 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
	 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include
	 /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks (framework directory)
	End of search list.
	conftest.c:99:2: error: implicit declaration of function 'at_quick_exit' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
	 at_quick_exit(func);
	 ^
	1 error generated.

Detected while diagnosing ticket  [#64550 _freeciv @2.6.6: at_quick_exit gives -Wimplicit-function-declaration warning, false positive_](https://trac.macports.org/ticket/64550 ).

#### Description

Apple clang version 11.0.0 (clang-1100.0.33.12) appears to lack the C11 language feature 'at_quick_exit' when compiling for macOS X 10.14. This causes the compiler to generate a `-Wimplicit-function-declaration` error message, even for correct configuration code, which correctly has a `#include <stdlib.h>` directive. MacPorts elevates this to the main.log. But it is a false positive. The configuration code is not implicitly declaring 'at_quick_exit'. So, this function name belongs in the whitelist. Because Apple rarely removes compiler features once introduced, it's a good guess that all macOS X versions before 10.4 (i.e. 10.4u through 10.13), also lack the 'at_quick_exit' feature. Thus this PR adds 'at_quick_exit' to whitelists for all of these macOS X versions.

###### Type(s)

- [X] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on
macOS 10.14.6 18G9323 x86_64
Xcode 11.2.1 11B500

###### Verification 
Have you

- [X] (sort of) followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [X] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

### CODE REVIEW

**Comment by JDLH** [comment]
Ryan wrote, in a [macports-dev list message on 2022-01-29](https://lists.macports.org/pipermail/macports-dev/2022-January/044099.html),

> … if you determine that a function is not in an SDK, it seems almost certain that it won't be in any earlier SDK either. Apple adds functions to the OS over time; it doesn't tend to remove them (at least when it comes to standard language features; they do on occasion remove Apple-specific functionality like certain frameworks).… 

Is that enough reason to add a similar `at_quick_exit` entry to each of `macosx10.4u.sdk.list` through `macosx10.13.sdk.list` (10 files)?

**Comment by kencu** [comment]
I haven’t checked which if any darwin versions have that function available yet, but I see there are available replacement functions that might be added to legacysupport, like this one:

https://git.stag-os.org/StagOS/android_system_core/commit/0c44256ae45968080c562aaf743f63019f580763

**Review by neverpanic — APPROVED** [review]
Yes, I would say that's enough reason to also add `at_quick_exit` to the older SDKs, yes.

Do you want to update this change to also cover this, or are you planning to do that in a separate change?

**Comment by JDLH** [comment]
> Yes, I would say that's enough reason to also add `at_quick_exit` to the older SDKs, yes.
> Do you want to update this change to also cover this, or are you planning to do that in a separate change?

Given that the existing change seems to not have merged yet, I will update this change.

I'm new to the MacPorts patch integration process. Should I worry that "some checks were not successful"?

**Comment by JDLH** [comment]
I have changed the other files macosx10.4u.sdk.list through 10.13 . I did this via the GitHub.com web UI. It created a separate commit for each file changed. I would like to squash these commits into one, but I can't figure out a way to do that via the GitHub.com web UI.  

1) should I squash the commits in this branch?

2) is there a way to do that via the GitHub.com web UI, or do I need to clone this repo on my client and do it on my client?

3) can I include the first commit, which was already requested for pulling, in my squash, but keep the same PR?  I worry that including the first commit in the squash means that a commit ID will get lost, and I believe that is bad. 

4) alternatively, should I make a separate branch and a new PR with the squashed commit for all macosx versions 10.4u through 10.14?

Thank you for helping me with the mechanics of GitHub. I don't see this explained in their documentation.

**Comment by neverpanic** [comment]
> I'm new to the MacPorts patch integration process. Should I worry that "some checks were not successful"?

No. That's a bug in our CI system, which does not expect a pull request that does not change any ports.

>  1. should I squash the commits in this branch?

Yes, please. However, I can create a squash merge, so it's probably quickest for us to just do that.

>  2. is there a way to do that via the GitHub.com web UI, or do I need to clone this repo on my client and do it on my client?

Not for you, no – there's the option on doing it with the merge button, but I'm not aware of any other GitHub UI feature to squash changes.

> 3. can I include the first commit, which was already requested for pulling, in my squash, but keep the same PR?  I worry that including the first commit in the squash means that a commit ID will get lost, and I believe that is bad.

Yes, you can include the first commit. The commit-ID will change, but that doesn't matter as long as we have not merged it.

> 4. alternatively, should I make a separate branch and a new PR with the squashed commit for all macosx versions 10.4u through 10.14?

Not necessary.

I'll squash-merge this now.

**Comment by neverpanic** [comment]
I've also reformatted the commit message (#numbers are claimed by github and will always link to pull requests, unfortunately, so we always reference trac tickets with absolute URL).

I've also added the `Closes: ` keyword so that the trac ticket will get auto-closed once this is merged.

**Comment by JDLH** [comment]
Thank you for your help, Clemens!

**Pull request merged**
